File: mac/src/Models/GameState.swift
import Foundation

public struct GameState: Codable, Equatable {
    public enum DrawMode: String, Codable {
        case drawOne
        case drawThree
    }
    
    public var stock: Pile
    public var waste: Pile
    public var foundations: [Pile] // Array of 4 foundations
    public var tableau: [Pile]     // Array of 7 columns
    
    public var score: Int
    public var movesCount: Int
    public var timerSeconds: Int
    public var isTimerActive: Bool
    public var drawMode: DrawMode
    public var hasWon: Bool
    public var recyclesCount: Int
    // Tracks how many cards from the current draw batch are still on top of the waste pile.
    // Shrinks as cards are played; reset to the draw count on each new draw.
    public var wasteDisplayCount: Int = 0

    public init(
        stock: Pile = Pile(id: "stock", type: .stock),
        waste: Pile = Pile(id: "waste", type: .waste),
        foundations: [Pile] = [],
        tableau: [Pile] = [],
        score: Int = 0,
        movesCount: Int = 0,
        timerSeconds: Int = 0,
        isTimerActive: Bool = false,
        drawMode: DrawMode = .drawThree,
        hasWon: Bool = false,
        recyclesCount: Int = 0,
        wasteDisplayCount: Int = 0
    ) {
        self.stock = stock
        self.waste = waste
        let foundationSuits: [Card.Suit] = [.spades, .clubs, .diamonds, .hearts]
        self.foundations = foundations.isEmpty ? foundationSuits.map { Pile(id: "foundation_\($0.rawValue)", type: .foundation) } : foundations
        self.tableau = tableau.isEmpty ? (0..<7).map { Pile(id: "tableau_\($0)", type: .tableau) } : tableau
        self.score = score
        self.movesCount = movesCount
        self.timerSeconds = timerSeconds
        self.isTimerActive = isTimerActive
        self.drawMode = drawMode
        self.hasWon = hasWon
        self.recyclesCount = recyclesCount
        self.wasteDisplayCount = wasteDisplayCount
    }
}

File: android/app/src/main/java/com/leah/honeycomb/klondike/GameState.kt
package com.leah.honeycomb.klondike

import com.leah.honeycomb.Card
import com.leah.honeycomb.Pile
import com.leah.honeycomb.PileType
import kotlinx.serialization.Serializable

@Serializable
enum class DrawMode {
    DrawOne,
    DrawThree
}

@Serializable
data class GameState(
    val stock: Pile = Pile(id = "stock", type = PileType.Stock),
    val waste: Pile = Pile(id = "waste", type = PileType.Waste),
    val foundations: List<Pile> = listOf(
        Pile(id = "foundation_Spades", type = PileType.Foundation),
        Pile(id = "foundation_Clubs", type = PileType.Foundation),
        Pile(id = "foundation_Diamonds", type = PileType.Foundation),
        Pile(id = "foundation_Hearts", type = PileType.Foundation)
    ),
    val tableau: List<Pile> = List(7) { index -> Pile(id = "tableau_$index", type = PileType.Tableau) },
    val score: Int = 0,
    val movesCount: Int = 0,
    val timerSeconds: Int = 0,
    val isTimerActive: Boolean = false,
    val drawMode: DrawMode = DrawMode.DrawThree,
    val hasWon: Boolean = false,
    val recyclesCount: Int = 0,
    val wasteDisplayCount: Int = 0,
    val vegasBankroll: Int = 0,
    val vegasBankrollAtGameStart: Int = 0
)
